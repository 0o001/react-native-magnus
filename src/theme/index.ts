export { ThemeContext } from './themeContext';
export { withTheme } from './withTheme';
export { ThemeProvider, ThemeProviderProps } from './themeProvider.component';

export {
  ThemeType,
  BorderPropsType,
  SpacingPropsType,
  RoundedPropsType,
} from './type';
