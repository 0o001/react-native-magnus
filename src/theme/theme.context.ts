import * as React from 'react';

import { ThemeType } from './type';

const defaultValue: ThemeType = {};

export interface ThemeContextType {
  theme: ThemeType;
  setTheme?: (theme?: ThemeType) => void;
}

export const THEME_CONTEXT_DEFAULT_VALUE = {
  theme: defaultValue,
};

export const ThemeContext: React.Context<ThemeContextType> = React.createContext(
  THEME_CONTEXT_DEFAULT_VALUE
);
