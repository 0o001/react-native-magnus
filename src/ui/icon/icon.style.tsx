import { StyleSheet } from "react-native";

import {
  getThemeProperty,
  createBorderWidthStyles,
  createBorderColorStyles,
  createSpacingStyles,
  createBorderRadiusStyles
} from "../../theme/theme.service";

/**
 * computed style
 *
 * @param theme
 * @param props
 */
export const getStyle = (theme: any, props: any) => {
  const computedStyle: any = {};

  computedStyle.container = {
    alignSelf: props.alignSelf,
    color: getThemeProperty(theme.colors, props.color, "#000"),
    fontSize: getThemeProperty(theme.fontSize, props.fontSize, 16),
    backgroundColor: getThemeProperty(theme.colors, props.bg, "transparent"),
    alignItems: "center",
    justifyContent: "center",
    ...createBorderWidthStyles(props),
    ...createSpacingStyles(props, theme.spacing),
    ...createBorderColorStyles(props, theme.colors),
    ...createBorderRadiusStyles(props, theme.borderRadius)
  };

  if (props.h) {
    computedStyle.container = {
      ...computedStyle.container,
      height: props.h
    };
  }

  if (props.w) {
    computedStyle.container = {
      ...computedStyle.container,
      width: props.w
    };
  }

  if (props.minH) {
    computedStyle.container = {
      ...computedStyle.container,
      minHeight: props.minH
    };
  }

  if (props.minW) {
    computedStyle.container = {
      ...computedStyle.container,
      minWidth: props.minW
    };
  }

  if (props.position) {
    computedStyle.container = {
      ...computedStyle.container,
      position: props.position
    };
  }

  if (typeof props.top !== "undefined") {
    computedStyle.container = {
      ...computedStyle.container,
      top: props.top
    };
  }

  if (typeof props.right !== "undefined") {
    computedStyle.container = {
      ...computedStyle.container,
      right: props.right
    };
  }

  if (typeof props.bottom !== "undefined") {
    computedStyle.container = {
      ...computedStyle.container,
      bottom: props.bottom
    };
  }

  if (typeof props.left !== "undefined") {
    computedStyle.container = {
      ...computedStyle.container,
      left: props.left
    };
  }

  if (props.shadow) {
    computedStyle.container = {
      ...computedStyle.container,
      ...theme.shadow[props.shadow],
      shadowColor: getThemeProperty(theme.colors, props.shadowColor, "white")
    };
  }

  // merging style props to computed style
  if (props.style) {
    computedStyle.container = {
      ...computedStyle.container,
      ...props.style
    };
  }

  return StyleSheet.create(computedStyle);
};
