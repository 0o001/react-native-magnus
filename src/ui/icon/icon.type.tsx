import { ViewProps as RNViewProps } from "react-native";

import {
  BorderPropsType,
  SpacingPropsType,
  RoundedPropsType
} from "../../theme";

export type iconFontFamilyType =
  | "AntDesign"
  | "Entypo"
  | "EvilIcons"
  | "Feather"
  | "FontAwesome"
  | "FontAwesome5"
  | "Foundation"
  | "Ionicons"
  | "MaterialIcons"
  | "MaterialCommunityIcons"
  | "Octicons"
  | "Zocial"
  | "SimpleLineIcons";

export interface IconProps
  extends RNViewProps,
    BorderPropsType,
    SpacingPropsType,
    RoundedPropsType {
  h?: number;
  w?: number;
  minW?: number;
  minH?: number;
  maxH?: number;
  maxW?: number;
  bg?: string;
  name: string;
  color?: string;
  shadow?: number;
  shadowColor?: string;
  position?: "absolute" | "relaitve";
  top?: string | number;
  left?: string | number;
  right?: string | number;
  bottom?: string | number;
  fontFamily?: iconFontFamilyType;
  fontSize?: string;
  zIndex?: number;
  opacity?: number;
  alignSelf?:
    | "auto"
    | "flex-start"
    | "flex-end"
    | "center"
    | "stretch"
    | "baseline";
}
