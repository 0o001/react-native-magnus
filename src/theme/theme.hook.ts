import React, { useState } from "react";

import { ThemeType } from "./type";
import { computeTheme } from "./theme.service";
import { ThemeContextType } from "./theme.context";

export const useTheme = (initialTheme?: ThemeType): ThemeContextType => {
  const [theme, setTheme] = useState(computeTheme(initialTheme));

  const setCurrentTheme = React.useCallback((newTheme?: ThemeType): void => {
    setTheme(computeTheme(newTheme));
  }, []);

  return {
    theme,
    setTheme: setCurrentTheme
  };
};
