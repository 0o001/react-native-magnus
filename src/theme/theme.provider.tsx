import * as React from "react";

import { ThemeType } from "./type";

import { ThemeContext } from "./theme.context";
import { useTheme } from "./theme.hook";

export interface ThemeProviderProps {
  theme?: ThemeType;
  children?: React.ReactNode;
}

export const ThemeProvider: React.FunctionComponent<ThemeProviderProps> = (
  props: ThemeProviderProps
) => {
  const { theme: themeProp, children } = props;
  const theme = useTheme(themeProp);

  return (
    <ThemeContext.Provider value={theme}>{children}</ThemeContext.Provider>
  );
};
