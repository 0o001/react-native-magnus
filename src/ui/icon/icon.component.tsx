import * as React from "react";
import { useContext } from "react";
import { View as RNView } from "react-native";

import { getStyle } from "./icon.style";
import { getIconSet } from "./icon.service";
import { ThemeContext } from "../../theme";
import { IconProps } from "./icon.type";
import { getThemeProperty } from "../../theme/theme.service";

const Icon: React.FunctionComponent<IconProps> = props => {
  const {
    m,
    mt,
    mr,
    mb,
    ml,
    p,
    pr,
    pt,
    pb,
    pl,
    h,
    w,
    bg,
    position,
    left,
    top,
    bottom,
    right,
    name,
    color,
    rounded,
    maxH,
    maxW,
    minW,
    minH,
    roundedTop,
    roundedRight,
    roundedBottom,
    roundedLeft,
    shadow,
    shadowColor,
    borderColor,
    borderBottomColor,
    borderLeftColor,
    borderTopColor,
    borderRightColor,
    borderWidth,
    borderLeftWidth,
    borderRightWidth,
    borderBottomWidth,
    borderTopWidth,
    fontFamily,
    fontSize,
    zIndex,
    style,
    ...rest
  } = props;
  const theme = useContext(ThemeContext);
  const computedStyle = getStyle(theme, props);

  const Icon = getIconSet(fontFamily);

  return (
    <RNView style={computedStyle.container} {...rest}>
      <Icon
        color={getThemeProperty(theme.colors, color, color)}
        size={getThemeProperty(theme.fontSize, fontSize, 16)}
        name={name}
        style={computedStyle.icon}
      />
    </RNView>
  );
};

Icon.defaultProps = {
  color: "gray500",
  fontSize: "text400",
  fontFamily: "AntDesign"
};

export { Icon };
