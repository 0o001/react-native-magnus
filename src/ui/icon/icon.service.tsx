import {
  AntDesign,
  Entypo,
  EvilIcons,
  Feather,
  FontAwesome,
  FontAwesome5,
  Fontisto,
  Foundation,
  Ionicons,
  MaterialCommunityIcons,
  MaterialIcons,
  Zocial,
  Octicons,
  SimpleLineIcons,
} from '@expo/vector-icons';

import { iconFontFamilyType } from './icon.type';

/**
 * get icon set
 */
export const getIconSet = (fontFamily: iconFontFamilyType = 'FontAwesome') => {
  if (fontFamily === 'AntDesign') {
    return AntDesign;
  }

  if (fontFamily === 'Entypo') {
    return Entypo;
  }

  if (fontFamily === 'EvilIcons') {
    return EvilIcons;
  }

  if (fontFamily === 'Feather') {
    return Feather;
  }

  if (fontFamily === 'FontAwesome') {
    return FontAwesome;
  }

  if (fontFamily === 'FontAwesome5') {
    return FontAwesome5;
  }

  if (fontFamily === 'Foundation') {
    return Foundation;
  }

  if (fontFamily === 'Ionicons') {
    return Ionicons;
  }

  if (fontFamily === 'MaterialIcons') {
    return MaterialIcons;
  }

  if (fontFamily === 'MaterialCommunityIcons') {
    return MaterialCommunityIcons;
  }

  if (fontFamily === 'Zocial') {
    return Zocial;
  }

  if (fontFamily === 'SimpleLineIcons') {
    return SimpleLineIcons;
  }

  if (fontFamily === 'Octicons') {
    return Octicons;
  }

  if (fontFamily === 'Fontisto') {
    return Fontisto;
  }

  return Ionicons;
};
