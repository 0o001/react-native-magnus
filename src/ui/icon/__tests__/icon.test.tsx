import React from "react";
import renderer from "react-test-renderer";

import { Icon } from "../icon.component";
import { ThemeProvider } from "../../../theme";

describe("Icon Component", () => {
  it("should render without issues", () => {
    const instance = renderer.create(
      <ThemeProvider>
        <Icon name="plus" />
      </ThemeProvider>
    );

    expect(instance.toJSON()).toMatchSnapshot();

    const typeInstance = instance.root.findByType(Icon);
    expect(typeInstance.props.name).toBe("plus");
  });
});
