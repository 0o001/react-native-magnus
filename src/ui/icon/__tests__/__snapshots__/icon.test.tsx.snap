// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon Component should render without issues 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "color": "#8c8c8c",
      "fontSize": 15,
      "justifyContent": "center",
    }
  }
>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": "#8c8c8c",
          "fontSize": 15,
        },
        undefined,
        Object {
          "fontFamily": "anticon",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    î˜§
  </Text>
</View>
`;
